Added zero-initialization to previously uninitialized HexValue structs

AntonJohansson opened the pull request.

Commit:
target/hexagon: import additional tests

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Niccolò Izzo <nizzo@rev.ng>

aleclearmind committed:
target/hexagon: call idef-parser functions

Extend gen_tcg_funcs.py in order to emit calls to the functions emitted
by the idef-parser, if available.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>

Commit:
target/hexagon: make helper functions non-static

Make certain helper functions non-static, making them available outside
genptr.c. These functions are required by code generated by the
idef-parser.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>
Reviewed-by: Richard Henderson <richard.henderson@linaro.org>

Commit:
target/hexagon: introduce new helper functions

These helpers will be employed by the idef-parser generated code.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Niccolò Izzo <nizzo@rev.ng>

Commit:
target/hexagon: expose next PC in DisasContext

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>
Reviewed-by: Richard Henderson <richard.henderson@linaro.org>

aleclearmind committed:
target/hexagon: prepare input for the idef-parser

Introduce infrastructure necessary to produce a file suitable for being
parsed by the idef-parser.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>

Commit:
target/hexagon: import lexer for idef-parser

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>

Commit:
target/hexagon: import parser for idef-parser

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>

aleclearmind committed:
target/hexagon: import README for idef-parser

Signed-off-by: Alessandro Di Federico <ale@rev.ng>

aleclearmind committed:
target/hexagon: update MAINTAINERS for idef-parser

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Reviewed-by: Richard Henderson <richard.henderson@linaro.org>

aleclearmind committed:
tcg: expose TCGCond manipulation routines

This commit moves into a separate file routines used to manipulate
TCGCond. These will be employed by the idef-parser.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>

Commit:
target/hexagon: make slot number an unsigned

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>
Acked-by: Richard Henderson <richard.henderson@linaro.org>

Commit:
target/hexagon: remove unused macros and functions

Signed-off-by: Alessandro Di Federico <ale@rev.ng>
Signed-off-by: Paolo Montesel <babush@rev.ng>

aleclearmind committed:
gitlab-ci: do not use qemu-project Docker registry

This commit is necessary in order to use container built by the current
run of the CI. If we don't do this, we use official containers which are
not affected by the additional dependencies we're introducing.

Signed-off-by: Alessandro Di Federico <ale@rev.ng>

Commit:
make slot uint32_t

Commit:
use fbrev helper

Commit:
slot is uint32_t

Commit:
i32 => tl

Commit:
ext => ovfl

Commit:
another slot

Commit:
remove unused token

Commit:
taylor unsigned patch

Commit:
unsigned

Commit:
refactor control registers

Commit:
fix fNEWREG

Commit:
fix SA/LC

Commit:
remove useless trunc/ext

Commit:
adjust store/load

Commit:
make integers default to 32 bit

Commit:
reintroduce constext

Commit:
remove QEMU_GENERIC

Commit:
remove unused include

Commit:
use tcg_gen_abs

Commit:
READ_REG => gen_read_reg

Commit:
propagate is_unsigned

Commit:
merge max/min

Commit:
assert size

Commit:
gen_rvalue_free

Commit:
reg_compose

Commit:
gen_rvalue_extend

Commit:
gen_rvalue_truncate

Commit:
gen_varid_allocate

Commit:
remove unused function

aleclearmind committed:
tmp

aleclearmind committed:
Build system stuff

aleclearmind committed:
Drop andp

aleclearmind committed:
Drop useless space

aleclearmind committed:
Drop some tests

AntonJohansson committed:
Added zero-initialization to previously uninitialized HexValue structs

aleclearmind committed:
Ensure we always set up signedness

AntonJohansson committed:
Changed strictness of C compiler

Now using `-Wall, -Wextra` which produced warnings about sign-compare
and unused parameters. These warnings then fixed. The other notable
change is the adding of `%option never-interactive` which removes
the dependency on the undefined `fileno` posix function which is used
when checking whether or not the parser is being run in a tty. This change
is not a problem as we don't use the interactive mode.

AntonJohansson committed:
Merge branch 'feature/for-upstream' of https://github.com/revng/qemu-hexagon into revng-feature/for-upstream

AntonJohansson committed:
Merge branch 'revng-feature/for-upstream' into feature/for-upstream

aleclearmind commented on target/hexagon/idef-parser/idef-parser.y:
```diff
@@ -934,7 +934,7 @@ int main(int argc, char **argv)
                               input_size,
                               input_file);
-    if (read_chars != input_size) {
+    if (read_chars != (size_t)input_size) {
```
```suggestion
    if (read_chars != (size_t) input_size) {
```

aleclearmind commented on target/hexagon/idef-parser/parser-helpers.c:
```diff
@@ -80,36 +80,43 @@ bool is_inside_ternary(Context *c)
 /* Print functions */
 void str_print(Context *c, YYLTYPE *locp, const char *string)
 {
+	(void)locp;
```
Mh, do any of the `_print` method ever uses `locp`? If not, let's drop the arguments everywhere.

aleclearmind replied:
Mind the whitespaces.
```suggestion
	  (void) locp;
```

AntonJohansson replied:
There are some print functions that use `locp`, namely `rvalue_out`, `imm_out`, and `reg_arg_print` so we can't drop it sadly. I've also gone ahead and made `reg_arg_print` static as it is not exposed in the header.

I have changed my editor settings to use 4 spaces as per the qemu coding standard so that should fix most indent problems. To clarify, should `(void) locp` be indented two spaces relative to the rest of the code? Haven't seen that before.

aleclearmind replied:
> There are some print functions that use `locp`, namely `rvalue_out`, `imm_out`, and `reg_arg_print` so we can't drop it sadly. I've also gone ahead and made `reg_arg_print` static as it is not exposed in the header.

ACK.

> I have changed my editor settings to use 4 spaces as per the qemu coding standard so that should fix most indent problems. To clarify, should `(void) locp` be indented two spaces relative to the rest of the code? Haven't seen that before.

Uhm, no, just like any other line. But here it was indented less then the next line.

aleclearmind commented on target/hexagon/idef-parser/parser-helpers.c:
```diff
@@ -1877,6 +1890,7 @@ void gen_load(Context *c, YYLTYPE *locp, HexValue *num, HexValue *size,
 void gen_store(Context *c, YYLTYPE *locp, HexValue *num, HexValue *size,
                HexValue *ea, HexValue *src)
 {
+	(void)num;
```
Please review if it makes sense to drop unused arguments.

AntonJohansson replied:
So `num` comes from the `fSTORE(NUM, ...)` and `fLOAD(NUM, ...)` macros but it's always set to `1`, is it for dealing with arrays? It might then be good to assert `num->imm.value == 1`.

aleclearmind replied:
Mmh, let's just drop it from `gen_store` and `assert` that in the caller. If in the future we ever needs this we can fix it.

AntonJohansson replied:
Sounds good

aleclearmind commented on target/hexagon/idef-parser/idef-parser.lex:
```diff
@@ -2,6 +2,7 @@
 %option 8bit reentrant bison-bridge
 %option warn nodefault
 %option bison-locations
+%option never-interactive
```
Why?

AntonJohansson replied:
Flex has the ability to be used interactively by default. This adds a call to `isatty(fileno(...))` to check whether or not it is used interactively. `fileno` is undeclared in C11 and so I added the `%option never-interactive` to get rid of the `isatty(fileno(...))` check.

I realize now that this necessary as QEMU uses `gnuC99` which includes `fileno`. I think its a good idea to remove the `C11` flag then and use `gnuC99` to comply with QEMU coding standards. This will also mean changing the C11 macro `_Generic` for their version `QEMU_GENERIC`.

aleclearmind replied:
> Flex has the ability to be used interactively by default. This adds a call to `isatty(fileno(...))` to check whether or not it is used interactively. `fileno` is undeclared in C11 and so I added the `%option never-interactive` to get rid of the `isatty(fileno(...))` check.

I see.

> I realize now that this necessary as QEMU uses `gnuC99` which includes `fileno`.

You mean "is *not* necessary"?

> I think its a good idea to remove the `C11` flag then and use `gnuC99` to comply with QEMU coding standards. This will also mean changing the C11 macro `_Generic` for their version `QEMU_GENERIC`.

Ideally we should not add flags such as `-std=c11` and use whatever is globally provided. That's why I asked the current command line in the other command.

:+1: for `QEMU_GENERIC`.

AntonJohansson replied:
> You mean "is not necessary"?

Oh yes, my bad.

aleclearmind commented on target/hexagon/meson.build:
```diff
@@ -177,6 +177,7 @@ idef_parser = executable('idef-parser',
                           idef_parser_dir / 'parser-helpers.c'],
                          include_directories: ['idef-parser', '../../include/'],
                          dependencies: [glib_dep],
+						 c_args: ['-std=c11', '-Wextra', '-Wall'],
```
Mind the indentation.
What's the current set of arguments?

You should be able to see it by:
```
ninja --verbose
```

AntonJohansson replied:
I'll remove the `C11` flag as I mentioned above to comply with coding standards, which means changing some macros around in `parser-helpers.h`. I'll remove '-Wall' as well since it is already included. However ´-Wextra' is not included by QEMU's build script, but I think it's good to keep for our code to get some better warnings.

aleclearmind commented:
Please never use `git merge`, always use `git rebase` (or if you're using `git pull` use `git pull --rebase`).

AntonJohansson commented:
> Please never use `git merge`, always use `git rebase` (or if you're using `git pull` use `git pull --rebase`).
ACK, I'll keep this in mind in the future!

AntonJohansson committed:
Changed back from `-std=c11` to `-std=gnu99`

Also removed duplicate flags such as `-Wall`. The change to `gnu99` is
done to comply with QEMU coding standards.

This resulted in the removal of `%option never-interactive` as `fileno`
is declared in `gnu99` which is not the case in `c11`.

Changing to `gnu99` also removes the `_Generic` macro from `c11`,
thankfully QEMU has their own `QEMU_GENERIC` macro to take its
place.

AntonJohansson committed:
Fixed indentation and removed some unused params

Indentation of `(void) ...` are now correct in the functions, and the
`reg_arg_print` function was made `static` as it is local to
`parser-helpers.c`. Lastly the `num` parameter is removed from the
`gen_load/store` functions as array arguments aren't handled, and
`num == 1` is asserted at the call sites in `idef-parser.y`.

aleclearmind commented on target/hexagon/idef-parser/parser-helpers.h:
```diff
@@ -87,14 +90,14 @@ void out_assert(Context *c, YYLTYPE *locp, void *dummy);
-        uint8_t:   uint8_print,                 \
-        uint64_t:  uint64_print,                \
-        int:       int_print,                   \
-        unsigned:  uint_print,                  \
-        HexValue:  rvalue_out,                  \
-        default:   out_assert                   \
+    QEMU_GENERIC(*x,                            \
+        (char,     str_print),                  \
+        (uint8_t,  uint8_print),                \
+        (uint64_t, uint64_print),               \
+        (int,      int_print),                  \
+        (unsigned, uint_print),                 \
+        (HexValue, rvalue_out),                 \
+        out_assert                              \
```
Actually, nope, drop this. I've just rebased on the latest QEMU and they have switched to C11:

```
$ git grep -w gnu11 | cat
configure:    CONFIGURE_CXXFLAGS=$(echo "$CONFIGURE_CFLAGS" | sed s/-std=gnu11/-std=gnu++11/)
configure:CONFIGURE_CFLAGS="-std=gnu11 -Wall"
meson.build:        default_options: ['warning_level=1', 'c_std=gnu11', 'cpp_std=gnu++11', 'b_colorout=auto'] +
```

Also, the macro `QEMU_GENERIC` no longer exists!

I'll soon push the branch rebased on today QEMU's `master`

AntonJohansson replied:
Haha alright!

AntonJohansson committed:
Changed back to using `_Generic`

As we'll be rebasing to more recent QEMU, which now uses `-std=gnu11`.

aleclearmind commented:
These changes have been merged in the v6 version of the patchset.
Thanks!
